In [1]:
import numpy as np
import matplotlib.pyplot as plt

Find the constants $x_0, x_1$ and $c$ such that the numerical integration formula $\int_0^2 f(x)dx = f(x_0)+cf(x_1)$ has the highest precision possible.

Obviously the answer to this has to depend on the function $f$: Consider that for $f(x) = x$ the optimal answers are $c=1, x_0 = 0$ and $x_2=2$, which give an exact answer (since this is a polynomial of degree less than three), but for $f(x) = \sin((\pi x)/2)$ these values give a terrible approximation. Since this question was worded ambiguously, I'll answer the following question (and if you wanted the us to use the function $f$ from question 2, then the answer is trivially $c =0, x_1 = 0, x_2 = 2$):

(1) Similar to what was done in class, what are the optimal $c, x_0, x_1$ for the case of $f$ a polynomial of degree lesss than 3 (because there are 3 total unknowns)

Suppose $f(x) = a_0 + a_1x + a_2x^2 + a_3 x^3$. Then we get

\begin{gather*} 
    \int_0^2 f(x)dx = a_0\int_0^2dx + a_1\int_0^2 xdx + a_2\int_0^2 x^2dx \\
    \text{which gives us the following system: } \{c = 2, \quad x_0 + cx_1 = 2, \quad x_0^2 + cx_1^2 = 8/3\}
\end{gather*}

In [6]:
#Note that this system of equations could have two solutions
#We'll have to find them solution and hope one of them lies inside [0,2] - otherwise I did something horribly wrong.
from sympy import *
x, y = symbols('x, y')
equation1 = Eq(x+2*y, 2)
equation2 = Eq(x**2 + 2*y**2, 8/3)
solutions = solve([equation1, equation2], [x, y])
print(solutions)

[(-0.276142374915397, 1.13807118745770), (1.60947570824873, 0.195262145875635)]


With this computation we get that the optimal values are:
\begin{gather*}
   c = 2 \\ x_0 = 1.60947570824873 \\ x_1 = 0.1952621245875635
\end{gather*}